extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -618,127 +601,154 @@ export class OpenPGPKey {
-      OpenPGPKey.getValidSigningKeyPacketFlags(key.keyPacket, primaryUser.selfCertification)
+      OpenPGPKey.getValidEncryptionKeyPacketFlags(key, primaryUser.user.selfCertifications[0]) | // todo: index?!
+      OpenPGPKey.getValidSigningKeyPacketFlags(key, primaryUser.user.selfCertifications[0]) // todo: index?!
     );
   };
 
   /**
    * Get latest self-signature date, in utc millis.
    * This is used to figure out how recently was key updated, and if one key is newer than other.
    */
-  private static getLastSigTime = async (key: OpenPGP.key.Key): Promise<number> => {
+  private static getLastSigTime = async (key: OpenPGP.Key): Promise<number> => {
     await key.getExpirationTime(); // will force all sigs to be verified
-    const allSignatures: OpenPGP.packet.Signature[] = [];
+    const allSignatures: OpenPGP.SignaturePacket[] = [];
     for (const user of key.users) {
       allSignatures.push(...user.selfCertifications);
     }
-    for (const subKey of key.subKeys) {
+    for (const subKey of key.subkeys) {
       allSignatures.push(...subKey.bindingSignatures);
     }
-    allSignatures.sort((a, b) => b.created.getTime() - a.created.getTime());
-    const newestSig = allSignatures.find(sig => sig.verified === true);
-    if (newestSig) {
-      return newestSig.created.getTime();
+    if (allSignatures.length > 0) {
+      // todo: all are verified now? .filter(x => x.verified)
+      return Math.max(...allSignatures.map(x => (x.created ? x.created.getTime() : 0)));
     }
     throw new Error('No valid signature found in key');
   };
 
-  private static extractExternalLibraryObjFromKey = (pubkey: Key) => {
-    if (pubkey.family !== 'openpgp') {
-      throw new UnexpectedKeyTypeError(`Key type is ${pubkey.family}, expecting OpenPGP`);
+  private static extractExternalLibraryObjFromKey = async (key: Key): Promise<OpenPGP.PrivateKey | OpenPGP.PublicKey> => {
+    if (key.family !== 'openpgp') {
+      throw new UnexpectedKeyTypeError(`Key type is ${key.family}, expecting OpenPGP`);
     }
-    const opgpKey = (pubkey as unknown as { internal: OpenPGP.key.Key }).internal;
-    if (!opgpKey) {
+    const keyWithPrivateFields = key as KeyWithPrivateFields;
+    const internal = keyWithPrivateFields.internal;
+    if (!internal) {
       throw new Error('Object has type == "openpgp" but no internal key.');
+    } else if (typeof internal === 'string') {
+      keyWithPrivateFields.internal = await opgp.readKey({ armoredKey: internal });
+      return keyWithPrivateFields.internal;
     }
-    return opgpKey;
+    return internal;
   };
 
-  private static extractStrengthUncheckedExternalLibraryObjFromKey = (pubkey: Key) => {
-    if (pubkey.family !== 'openpgp') {
-      throw new UnexpectedKeyTypeError(`Key type is ${pubkey.family}, expecting OpenPGP`);
+  private static extractStrengthUncheckedExternalLibraryObjFromKey = async (key: Key) => {
```

Comment:
Also these "StrengthUnchecked" etc checks could be removed if the above comment is true. Doesn't have to be now.

Reply:
#4905